wallet/api/spam/pow.go

```diff
@@ -78,53 +87,42 @@ func (t *txCounter) resize(n int64) {
 	t.store = newStore
 }
 
-type ProofOfWork struct {
-	// chainID to the counter for transactions sent.
-	counters map[string]*txCounter
-	mu       sync.Mutex
-}
-
-func NewProofOfWork() *ProofOfWork {
-	return &ProofOfWork{
-		counters: map[string]*txCounter{},
-	}
-}
-
-func (p *ProofOfWork) getCounterForChain(chainID string) *txCounter {
-	if _, ok := p.counters[chainID]; !ok {
-		p.counters[chainID] = &txCounter{}
+func (s *Handler) getCounterForChain(chainID string) *txCounter {
+	if _, ok := s.counters[chainID]; !ok {
+		s.counters[chainID] = &txCounter{}
 	}
-	return p.counters[chainID]
+	return s.counters[chainID]
 }
 
 // Generate returns a proof-of-work with difficult that respects the history of transactions sent in against a particular block.
-func (p *ProofOfWork) Generate(pubKey string, lastBlock *nodetypes.LastBlock) (*commandspb.ProofOfWork, error) {
-	p.mu.Lock()
-	counter := p.getCounterForChain(lastBlock.ChainID)
+func (s *Handler) GenerateProofOfWork(pubKey string, st *nodetypes.SpamStatistics) (*commandspb.ProofOfWork, error) {
+	s.mu.Lock()
+	counter := s.getCounterForChain(st.ChainID)
+	blockState := st.PoW.PowBlockStates[0]
```

Comment:
is this the last block or some random block?

Reply:
Should be the latest, we sort them in `insecure_grpc_adapter.go`.

Reply:
I wonder if we want to do something smarter, like try to find the block that minimises the difficulty but anyways, the last block would be fine most of the time (it won't if suddenly there's an extra long block)